Create Whats new in C# 7.1

BillWagner opened the pull request:
Fixes #2794

Note that this PR is dependent on PR #2835

/cc @Pilchie @jcouv @DustinCampbell

BillWagner committed:
interim checkin

BillWagner committed:
First draft completed.

BillWagner committed:
proofread and update

DustinCampbell commented on docs/csharp/whats-new/index.md:
```diff
@@ -21,9 +21,12 @@ major features added in each release.
 
-* [C# 7](csharp-7.md):
+* [C# 7.1](csharp-7-1.md):
     - This page describes the latest features in the C# language. This covers C# 7, currently available in [Visual Studio 2017](https://www.visualstudio.com/vs/whatsnew/).
```
"This covers C# 7" -> "This covers C# 7.1" ?

BillWagner replied:
Also added .NET Core 2.0

DustinCampbell commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+each features is ready. This release also adds the ability to configure
+the compiler to match a specified version of the language. This feature
+separates the decision to upgrade tools from the decision to upgrade language
+versions.
```
Not sure if you're interested in feedback on the actually copy, but the last four sentences in this paragraph all start with "This", which reads a bit awkwardly.

DustinCampbell commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+separates the decision to upgrade tools from the decision to upgrade language
+versions.
+
+This release adds the [language version selectionn](#language-version-selection)
```
And again. :smile:

DustinCampbell commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+The new language features in this release are:
+
+* [`async` `Main` method](#async-main)
+  - The entry point for a console application can have the `async` modifier.
```
`async Main` is not restricted to console applications.

DustinCampbell commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+* [`async` `Main` method](#async-main)
+  - The entry point for a console application can have the `async` modifier.
+* [`default` value expressions](#default-value-expressions)
+  - The `default(T)` expression may omit the target type when that type can be inferred.
```
Is this the correct name of the feature? I was under the impression they were called "default literal expressions"

BillWagner replied:
Right. I re-worked the sentence a bit. Yes, this new feature is "default literal expressions", which apply in the context of "default value expressions". I'm linking to updates to the topic on "default value expressions". Let me know what you think of the revision

DustinCampbell replied:
I think it's fine. It _is_ a bit weird to see "default value expressions" in a list of new features.

DustinCampbell commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+'Properties'. Select the "Build" tab and click on the "Advanced" button. In the dropdown,
+select "C# latest minor version (latest)", or the specific version "C# 7.1"
+as shwon in the image below. The 'latest' value means you want to use the latest
+minor versio on the current machine. The 'C# 7.1' means that you want to use C# 7.1,
```
"version"

DustinCampbell commented:
Thanks for including me @BillWagner. I did a quick read-through and left some feedback.

abelbraaksma commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+In Visual Studio, right-click on the project node in Solution explorer and select
+'Properties'. Select the "Build" tab and click on the "Advanced" button. In the dropdown,
+select "C# latest minor version (latest)", or the specific version "C# 7.1"
+as shwon in the image below. The 'latest' value means you want to use the latest
```
shwon > shown

abelbraaksma commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+C# 7.1 is the first point release to the C# Language. This release is the
+first step toward the next major release of the langauge. This increased
+cadence enables you to use the new features sooner, ideally when
+each features is ready. This release also adds the ability to configure
```
each features > each feature

rpetrusha commented on docs/csharp/whats-new/csharp-7-1.md:
````diff
@@ -0,0 +1,176 @@
+following lines:
+
+```xml
+<PropertyGroup>
````
I think it might be useful to include a full <PropertyGroup> element including the Configuration attribute. That helps to make it clear that this is going to be a per-configuration setting.

rpetrusha commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+> creates separate nodes for each build configuration. You'll typically
+> set the value the same in all build configurations.
+
+Valid settings for "langVersion" element are:
```
"langVersion" --> `LangVersion`

rpetrusha commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+
+> [!NOTE]
+> If you use the Visual Studio IDE to update your csproj files, the IDE
+> creates separate nodes for each build configuration. You'll typically
```
I think that the important thing to stress here is not that it's typically the same, but that you need to also manually do it for each.

BillWagner replied:
Let me know what you think of the updated version.

rpetrusha commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+## Async main
+
+An *async main* method enables you to use `await` in your `Main` method.
+Where previously you would need to write:
```
Where previously --> Previously

rpetrusha commented on docs/csharp/whats-new/csharp-7-1.md:
````diff
@@ -0,0 +1,176 @@
+}
+```
+
+If your program does not return an exit code, you can declare `Main` method
````
`Main` method --> a `Main` method

rpetrusha commented on docs/csharp/whats-new/csharp-7-1.md:
````diff
@@ -0,0 +1,176 @@
+```
+
+If your program does not return an exit code, you can declare `Main` method
+that returns a `Task`:
````
perhaps use an `<xref>` instead of fenced Task?

rpetrusha commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+There are many language constructs where default value expressions can make use
+of this new syntax. You can learn more in the topic on default value expressions
+in the topics on the [`default` keyword](../language-reference/keywords/default.md)
+and the topic on [default value expressions](../programming-guide/statements-expressions-operators/default-value-expressions.md).
```
There's a lot of repetition in this paragraph, starting with an unspecified topic on default value expressions, with (plural) topics and a link to a single topic on the default keyword, and then the reference to default value expressions.

BillWagner replied:
I rewrote the paragraph. Let me know what you think.

rpetrusha commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+## Inferred tuple element names
+
+This feature is a small enhancement to the tuples feature introduced in
+C# 7.0. Many times that you initialize a tuple, the variables used for the
```
Many times that --> Many times when

rpetrusha commented on docs/csharp/whats-new/csharp-7-1.md:
````diff
@@ -0,0 +1,176 @@
+var pair = (count, label); // element names are "count" and "label"
+```
+
+You can learn more about the details for this feature in the [Tuples](../tuples.md)) topic.
````
about the details for this feature --> about this feature

rpetrusha commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+
+## Reference assembly generation
+
+There are two new compiler options that generate *reference only assemblies*:
```
There are two new compiler options that --> Two new compiler options
reference only --> reference-only

BillWagner commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,176 @@
+
+## Langauge version selection
+
+The C# compiler that ships with Visual Studio 2017 V15.3 or with
```
I looked in the about box in Visual Studio. "Version" is spelled out. Updated.

BillWagner committed:
Updated per feeeback.

DustinCampbell reviewed (approved):
LGTM from my perspective

mairaw commented on docs/csharp/whats-new/csharp-7-1.md:
````diff
@@ -0,0 +1,187 @@
+Func<string, bool> whereClause = default(Func<string, bool>);
+```
+
+You can now omit the type on the right hand side of the initialization:
````
right-hand

mairaw commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,187 @@
+## Default literal expressions
+
+Default literal expressions are an enhancement to default value expressions.
+These initialize a variable to the default value. Where you previously
```
These -> These expressions

mairaw commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,187 @@
+
+## Language version selection
+
+The C# compiler that ships with Visual Studio 2017 version 15.3 or with
```
not sure if you wanna be future-proof here and say "and later versions" for the VS and SDK statements or "Starting with..."

mairaw commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,187 @@
+
+The C# compiler that ships with Visual Studio 2017 version 15.3 or with
+the .NET Core SDK 2.0 supports C# 7.1. However, the 7.1 features are turned
+off by default. To enable 7.1 features, you need to change the language
```
To enable *the* ?

mairaw commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,187 @@
+version setting for your project.
+
+In Visual Studio, right-click on the project node in Solution Explorer and select
+'Properties'. Select the "Build" tab and select the "Advanced" button. In the dropdown,
```
bold for UI options

mairaw commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,187 @@
+In Visual Studio, right-click on the project node in Solution Explorer and select
+'Properties'. Select the "Build" tab and select the "Advanced" button. In the dropdown,
+select "C# latest minor version (latest)", or the specific version "C# 7.1"
+as shown in the image below. The 'latest' value means you want to use the latest
```
below -> following

inline code latest so it doesn't get translated?

mairaw commented on docs/csharp/whats-new/csharp-7-1.md:
````diff
@@ -0,0 +1,187 @@
+</PropertyGroup>
+```
+
+Valid settings for "langVersion" element are:
````
langVersion -> LangVersion? inline code?

rpetrusha replied:
"langVersion" --> the `LangVersion`

mairaw commented on docs/csharp/whats-new/csharp-7-1.md:
````diff
@@ -0,0 +1,187 @@
+}
+```
+
+You can know write:
````
know -> now

mairaw commented on docs/csharp/whats-new/csharp-7-1.md:
````diff
@@ -0,0 +1,187 @@
+}
+```
+
+If your program does not return an exit code, you can declare a `Main` method
````
nit: does not -> doesn't

mairaw commented on docs/csharp/whats-new/csharp-7-1.md:
````diff
@@ -0,0 +1,187 @@
+```
+
+If your program does not return an exit code, you can declare a `Main` method
+that returns a @System.Threading.Tasks.Task:
````
should we standardize on always using `<xref:uid>`?

BillWagner replied:
That works. I wasn't sure which was preferred.

mairaw reviewed (changes requested):
Looks great! I have a couple of minor edits suggestions and comments.

rpetrusha reviewed (commented):
Along with Maira's comments, I've added one. Otherwise, you've done a good a good job of addressing my previous comments. This is a really good overview, @BillWagner.

BillWagner committed:
update per feedback.

BillWagner commented:
@rpetrusha @mairaw Thanks for the 🦅 👀 review.  I addressed all your comments.  Let me know what you think.

rpetrusha commented on docs/csharp/whats-new/csharp-7-1.md:
````diff
@@ -75,20 +75,20 @@ following lines:
 ```
 
-Valid settings for "langVersion" element are:
+Valid settings for `LangVersion` element are:
````
nit: the `LangVersion` element

rpetrusha reviewed (approved):
@BillWagner, except for one small nit, LGTM.

BillWagner committed:
one quick grammar fix

jcouv reviewed (approved):
LGTM
FYI @gafter @jaredpar

jcouv commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,187 @@
+  - You can use default literal expressions in default value expressions when the target type can be inferred.
+* [Inferred tuple element names](#inferred-tuple-element-names)
+  - The names of tuple elements can be inferred from tuple initialization in many cases.
+* [Reference assembly generation](#reference-assembly-generation)
```
Technically that's not a language feature (it's just a compiler feature), but I don't think it's worthwhile to change.

mairaw commented:
I'll review again but builds has warnings that need to be addressed (e.g. broken image links)

mairaw commented on docs/csharp/whats-new/csharp-7-1.md:
```diff
@@ -0,0 +1,187 @@
+You can learn more about this enhancement in the C# Programming Guide topic
+on [default value expressions](../programming-guide/statements-expressions-operators/default-value-expressions.md).
+
+This enhancement also changes some of the parsing rules for the [`default` keyword](../language-reference/keywords/default.md)
```
nit: this link behaves kind of weird because of the inline code in the middle. 
![image](https://user-images.githubusercontent.com/12971179/29696652-2838ea98-88ff-11e7-99d8-37514ab2775d.png)

also this is missing punctuation.

mairaw reviewed (approved):
Looks good overall. I've left a nit comment plus the new topic is missing from the TOC. After that, it's ready to merge.

BillWagner committed:
respond to review feedback.

BillWagner committed:
clarify language vs. compiler feature.

BillWagner commented:
Made final updates. I'll merge once #2835 is merged, and the build passes when those topics are live.

rpetrusha commented:
@BillWagner, the build completed with 3 warnings for the Getting Started topic, but for some reason the PR didn't register the build completion. I'm closing and reopening the PR.

BillWagner commented:
@rpetrusha Some of the warnings will be fixed when #2835 is merged. I've addressed all the feedback on that PR. Can you approve that one and I'll merge it?

rpetrusha commented:
@BillWagner, I approved it a while ago. We're just waiting on the build to complete -- the 3 warnings that I noted were from the OPS build report, but the build and this PR weren't in sync for some reason, and this PR showed the build as still in progress. I just restarted the build a few minutes ago, so once it's done, you should be able to merge.

BillWagner commented:
build is stuck. closing and reopening. 😭
